### ISSUE TITLE
override dual-stack: false on the server side

### ISSUE THREAD

**Issue by zvada** [xrootd/xrootd#411]
Hello!

We're experiencing issues with older clients vs redirector IPv6 enabled where client host is dual-stack. Relevant piece of behavior is from the log as follows:

`[2016-09-02 19:40:10.529888 +0200][Debug  ][XRootDTransport   ] [cmsxrootd.fnal.gov:1094 #0.0] Sending out kXR_login request, username: zvada, cgi: ?xrd.cc=ch&xrd.tz=1&xrd.appname=xrdcp&xrd.info=, dual-stack: false`
...
...

```
[2016-09-02 19:41:00.221320 +0200][Debug  ][File              ] [0x2000000@root://cmsxrootd.fnal.gov:1094//store/mc/RunIISpring16reHLT80/DYJetsToLL_M-50_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/AODSIM/PUSpring16RAWAODSIM_reHLT_80X_mcRun2_asymptotic_v14-v1/40000/003A70D5-E55C-E611-9B14-0025905C53A8.root] Error while opening at cms-xrd-transit.cern.ch:1094: [ERROR] Server responded with an error: [3011] No servers are available to read the file.
[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Server responded with an error: [3011] No servers are available to read the file.
$ xrdcp -version
v4.0.4
```

Forcing XRD_NETWORKSTACK=IPv4 solve the issue on client side, though:

```
$ XRD_NETWORKSTACK=IPv4 xrdcp -f root://cmsxrootd.fnal.gov//store/mc/RunIISpring16reHLT80/DYJetsToLL_M-50_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/AODSIM/PUSpring16RAWAODSIM_reHLT_80X_mcRun2_asymptotic_v14-v1/40000/003A70D5-E55C-E611-9B14-0025905C53A8.root /dev/null 
[80MB/3.495GB][  2%][=>                                                ][8.889MB/s] 
$ xrdcp -version
v4.0.4
```

Suggestion is provide on the server side a configuration parameter to force dual-stack to true - we'd assume that IPv4 is always present. Clients >=v4.3.x handle dual-stack situation well, but for users running anything below that would be preferable be cautions and act on the server side with config tweak if possible.

Thanks,
Marian

**Comment by abh3** [comment]
Hi Marian,

I see that the cluster configuration for cmsxrootd.fnal.gov is:

0: [::ffff:129.93.239.131] xrootd.unl.edu v4.3.0
1: [::ffff:131.225.204.174] cmsxrootd1.fnal.gov v4.3.0-rc3
2: [2600:900:6:1101:5054:ff:fe00:70cb] xrootd.unl.edu v4.3.0

Any particular reason why the fnal one is not on a supported version of 
xrootd? In any case, I will look to see if we addressed this issue in 
4.4.0. In the mean time, could you do:

xrdfs cmsxrootd.fnal.gov:1094 locate -h <the_problem_path>

and return the result to me? I can't do it as it appears you have turned 
on access control for the locate command.

Andy

B.T.W Version 4.4.0 is much better than 4.3.0 so you should consider 
upgrading.

**Comment by zvada** [comment]
Hi Andy,

On 9/2/16 2:27 PM, Andrew Hanushevsky wrote:
> Hi Marian,
>
> I see that the cluster configuration for cmsxrootd.fnal.gov is:
>
> 0: [::ffff:129.93.239.131] xrootd.unl.edu v4.3.0
> 1: [::ffff:131.225.204.174] cmsxrootd1.fnal.gov v4.3.0-rc3
> 2: [2600:900:6:1101:5054:ff:fe00:70cb] xrootd.unl.edu v4.3.0

which goofer you used get this info? Let me learn something here, please.

This is actually problem now with global redirectors, too. They have v6 
active interface and everyone using CMSSW now against global redirecotr 
cms-xrd-global.cern.ch hit this. We're not sure whether this is some 
coincidence or not with lxplus, but it's very weird because global 
redirector is on dual-stack for long time and noone noticed any issues 
until couple days ago when we put in EU regional level of redirection 
dual-stack host (xrootd-cms-redir-01.cr.cnaf.infn.it). Today we've taken 
it out, but problem for users is still there for this particular 
"dual-stack: false" reason.

> Any particular reason why the fnal one is not on a supported version of
> xrootd? In any case, I will look to see if we addressed this issue in
> 4.4.0.

not in OSG repo yet, just created OSG SW JIRA ticket update package to 
4.4.0. FNAL is still in -rc3, yeah, I need to ping them. Let me know if 
you find this fixed in 4.4.0, that'll be great.

> In the mean time, could you do:
>
> xrdfs cmsxrootd.fnal.gov:1094 locate -h <the_problem_path>
>
> and return the result to me? I can't do it as it appears you have turned
> on access control for the locate command.

looks like I'm blocked, too (with my proxy):
$ xrdfs cmsxrootd.fnal.gov:1094 locate -h 
/store/mc/RunIISpring16reHLT80/DYJetsToLL_M-50_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/AODSIM/PUSpring16RAWAODSIM_reHLT_80X_mcRun2_asymptotic_v14-v1/40000/003A70D5-E55C-E611-9B14-0025905C53A8.root 

[ERROR] Server responded with an error: [3010] Unable to locate 
/store/mc/RunIISpring16reHLT80/DYJetsToLL_M-50_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/AODSIM/PUSpring16RAWAODSIM_reHLT_80X_mcRun2_asymptotic_v14-v1/40000/003A70D5-E55C-E611-9B14-0025905C53A8.root; 
permission denied

Where is this control configurable, on which level and what knobs, please?

Thanks,
Marian

**Comment by xrootd-dev** [comment]
Hi Marian,

**Comment by zvada** [comment]
Btw, you might have good point about VM and v6 networking on it. We already had sort of vague suspect with PIC about this, that for instance they have one interface with an IPv6 and another one with IPv4 - they claim technically it’s not a dual-stack server but it has both stacks regardless, sort of design problem of WNs on vlan?
Though, I have no idea if this might be related somewhat to lxplus, but what we noticed just last week make me believe something must have changed on the lxplus pool which didnt' cause any sort of such issues we see now with dual-stack: false...

**Comment by bbockelm** [comment]
Hi,

While the discussion of DNS configuration is nice, I can we steer back to the underlying cause:

1. Old versions of the Xrootd client are known to log in to a server via IPv6 but set `dual-stack:false` despite the worker node actually being dual-stack.  This has already been fixed in newer Xrootd versions.
   * It is not an option to remove the old versions of the client from use until about 2019-or-so.
2. Current server versions adhere to protocol spec and trust the client's setting of the dual-stack flag.
3. We would like the ability to break the protocol spec and have the server always treat things as if IPv6 client's dual-stack flag was always set to true.

As far as I can tell, neither DNS configuration nor server version comes into play here.

Brian

**Comment by duncanrand** [comment]
Hi Marian

I would like to point out that in the future we may have IPv6-only worker nodes and so we shouldn't adopt a solution that assumes IPv4 is always present. In fact we are testing out an IPv6-only worker node at Brunel at the moment.

Duncan

**Comment by xrootd-dev** [comment]
Yes, I agree. One or our way-back original designs assumed 
clients would be dual-stack and we got a lot of push-back on that so the 
current implementation does not assume that. Do consider that we are only 
talking about clients here -- not servers and only old clients for that 
matter which don't really tell you much on their capabilities and when 
they do they are not always correct. So, if we restric it to this 
particulat case (and make it configurable on the server side) would that 
address your concerns?

Andy

**Comment by bbockelm** [comment]
@duncanrand - indeed, we haven't forgotten about the IPv6-only case.

However, the options here are all bad:
- Disable IPv6 completely.  Very undesirable.
- Disable IPv6 until buggy client is no longer used (2019).  Very undesirable.
- Break protocol to assume IPv4 connectivity is available.  Undesirable.
- Break protocol until buggy client is no longer used (2019), then switch off protocol break and fix IPv6-only clients.  Slightly less undesirable.

So, we'll have things correct once the oldest clients are cycled out.  I just feel that breaking IPv6-only cases is a better alternate than turning off IPv6 completely.

Andy - is there any other trick to identify old clients that are connecting to the server?  I didn't see a version string passed to the `kXR_login` command.

**Comment by ljanyst** [comment]
The client sends `kXR_protocol` piggy-backed after the initial handshake. https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClXRootDTransport.cc#L1153

**Comment by bbockelm** [comment]
I'm not particularly knowledgeable about the wire format, but poking around, it seems both the broken (4.0.4) and working client (4.2.x) use the same protocol version in `kXR_protocol` (preventing me from distinguishing them).

Or am I missing something?

**Comment by ljanyst** [comment]
I did not read the thread. After reading only your last message I figured you want to make changes to the protocol and need to find a way to distinguish old and new clients.

**Comment by bbockelm** [comment]
Ah - sorry, no.  The protocol design is correct: bugs in an old client version are causing operational issues.

**Comment by abh3** [comment]
Where exactly is the breakpoint? That is, client version x vs version y. I 
need to know that to see if we can actually distinguish between those 
versions.
Andy

**Comment by zvada** [comment]
We've found issue in the client v4.0.x ... not an issue from 4.3.0 and above.

**Comment by abh3** [comment]
We have pushed an option to do this and will be available in 4.6.
